File: easypermissions/src/main/java/com/sunplus/easypermissions/PermissionRequest.java
package com.sunplus.easypermissions;

import android.app.Activity;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;
import android.support.annotation.Size;
import android.support.annotation.StringRes;
import android.support.annotation.StyleRes;
import android.support.v4.app.Fragment;
import com.sunplus.easypermissions.helper.PermissionHelper;
import java.util.Arrays;

/**
 * Created by w.feng on 2018/10/10
 * Email: fengweisb@gmail.com
 */
public final class PermissionRequest {

  private final PermissionHelper mHelper;
  private final String[] mPerms;
  private final int mRequestCode;
  private final String mRationale;
  private final String mPositiveButtonText;
  private final String mNegativeButtonText;
  private final int mTheme;

  private PermissionRequest(PermissionHelper helper,
                            String[] perms,
                            int requestCode,
                            String rationale,
                            String positiveButtonText,
                            String negativeButtonText,
                            int theme) {
    mHelper = helper;
    mPerms = perms.clone();
    mRequestCode = requestCode;
    mRationale = rationale;
    mPositiveButtonText = positiveButtonText;
    mNegativeButtonText = negativeButtonText;
    mTheme = theme;
  }

  @NonNull
  @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
  public PermissionHelper getHelper() {
    return mHelper;
  }

  @NonNull
  public String[] getPerms() {
    return mPerms.clone();
  }

  public int getRequestCode() {
    return mRequestCode;
  }

  @NonNull
  public String getRationale() {
    return mRationale;
  }

  @NonNull
  public String getPositiveButtonText() {
    return mPositiveButtonText;
  }

  @NonNull
  public String getNegativeButtonText() {
    return mNegativeButtonText;
  }

  @StyleRes
  public int getTheme() {
    return mTheme;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    PermissionRequest request = (PermissionRequest) o;

    return Arrays.equals(mPerms, request.mPerms) && mRequestCode == request.mRequestCode;
  }

  @Override
  public int hashCode() {
    int result = Arrays.hashCode(mPerms);
    result = 31 * result + mRequestCode;
    return result;
  }

  @Override
  public String toString() {
    return "PermissionRequest{" +
        "mHelper=" + mHelper +
        ", mPerms=" + Arrays.toString(mPerms) +
        ", mRequestCode=" + mRequestCode +
        ", mRationale='" + mRationale + '\'' +
        ", mPositiveButtonText='" + mPositiveButtonText + '\'' +
        ", mNegativeButtonText='" + mNegativeButtonText + '\'' +
        ", mTheme=" + mTheme +
        '}';
  }

  public static final class Builder {
    private final PermissionHelper mHelper;
    private final int mRequestCode;
    private final String[] mPerms;

    private String mRationale;
    private String mPositiveButtonText;
    private String mNegativeButtonText;
    private int mTheme = -1;

    public Builder(@NonNull Activity activity, int requestCode,
                   @NonNull @Size(min = 1) String... perms) {
      mHelper = PermissionHelper.newInstance(activity);
      mRequestCode = requestCode;
      mPerms = perms;
    }

    public Builder(@NonNull Fragment fragment, int requestCode,
                   @NonNull @Size(min = 1) String... perms) {
      mHelper = PermissionHelper.newInstance(fragment);
      mRequestCode = requestCode;
      mPerms = perms;
    }

    @NonNull
    public Builder setRationale(@Nullable String rationale) {
      mRationale = rationale;
      return this;
    }

    @NonNull
    public Builder setRationale(@StringRes int resId) {
      mRationale = mHelper.getContext().getString(resId);
      return this;
    }

    @NonNull
    public Builder setPositiveButtonText(@Nullable String positiveButtonText) {
      mPositiveButtonText = positiveButtonText;
      return this;
    }

    @NonNull
    public Builder setPositiveButtonText(@StringRes int resId) {
      mPositiveButtonText = mHelper.getContext().getString(resId);
      return this;
    }

    @NonNull
    public Builder setNegativeButtonText(@Nullable String negativeButtonText) {
      mNegativeButtonText = negativeButtonText;
      return this;
    }

    @NonNull
    public Builder setNegativeButtonText(@StringRes int resId) {
      mNegativeButtonText = mHelper.getContext().getString(resId);
      return this;
    }

    @NonNull
    public Builder setTheme(@StyleRes int theme) {
      mTheme = theme;
      return this;
    }

    @NonNull
    public PermissionRequest build() {
      if (mRationale == null) {
        mRationale = mHelper.getContext().getString(R.string.rationale_ask);
      }
      if (mPositiveButtonText == null) {
        mPositiveButtonText = mHelper.getContext().getString(android.R.string.ok);
      }
      if (mNegativeButtonText == null) {
        mNegativeButtonText = mHelper.getContext().getString(android.R.string.cancel);
      }

      return new PermissionRequest(
          mHelper,
          mPerms,
          mRequestCode,
          mRationale,
          mPositiveButtonText,
          mNegativeButtonText,
          mTheme);
    }
  }
}

File: easypermissions/src/main/java/com/almosting/easypermissions/PermissionRequest.kt
package com.almosting.easypermissions

import android.R
import android.app.Activity
import androidx.annotation.RestrictTo
import androidx.annotation.RestrictTo.Scope.LIBRARY_GROUP
import androidx.annotation.Size
import androidx.annotation.StringRes
import androidx.annotation.StyleRes
import androidx.fragment.app.Fragment
import com.almosting.easypermissions.R.string
import com.almosting.easypermissions.helper.PermissionHelper
import java.util.Arrays

/**
 * Created by w.feng on 2018/10/10
 * Email: fengweisb@gmail.com
 */
class PermissionRequest private constructor(
  @get:RestrictTo(LIBRARY_GROUP) val helper: PermissionHelper<*>,
  perms: Array<String>,
  requestCode: Int,
  rationale: String,
  positiveButtonText: String,
  negativeButtonText: String,
  theme: Int
) {
  private val mPerms: Array<String>
  val requestCode: Int
  val rationale: String
  val positiveButtonText: String
  val negativeButtonText: String
  @get:StyleRes val theme: Int

  val perms: Array<String>
    get() = mPerms.clone()

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val request =
      other as PermissionRequest
    return Arrays.equals(mPerms, request.mPerms) && requestCode == request.requestCode
  }

  override fun hashCode(): Int {
    var result = Arrays.hashCode(mPerms)
    result = 31 * result + requestCode
    return result
  }

  override fun toString(): String {
    return "PermissionRequest{" +
        "mHelper=" + helper +
        ", mPerms=" + Arrays.toString(mPerms) +
        ", mRequestCode=" + requestCode +
        ", mRationale='" + rationale + '\'' +
        ", mPositiveButtonText='" + positiveButtonText + '\'' +
        ", mNegativeButtonText='" + negativeButtonText + '\'' +
        ", mTheme=" + theme +
        '}'
  }

  class Builder {
    private val mHelper: PermissionHelper<*>
    private val mRequestCode: Int
    private val mPerms: Array<String>
    private var mRationale: String? = null
    private var mPositiveButtonText: String? = null
    private var mNegativeButtonText: String? = null
    private var mTheme = -1

    constructor(
      activity: Activity, requestCode: Int,
      @Size(min = 1) vararg perms: String
    ) {
      mHelper =
        PermissionHelper.newInstance(activity)
      mRequestCode = requestCode
      mPerms = perms as Array<String>
    }

    constructor(
      fragment: Fragment, requestCode: Int,
      @Size(min = 1) vararg perms: String
    ) {
      mHelper =
        PermissionHelper.Companion.newInstance(fragment)
      mRequestCode = requestCode
      mPerms = perms as Array<String>
    }

    fun setRationale(rationale: String?): Builder {
      mRationale = rationale
      return this
    }

    fun setRationale(
      @StringRes resId: Int
    ): Builder {
      mRationale = mHelper.context!!.getString(resId)
      return this
    }

    fun setPositiveButtonText(positiveButtonText: String?): Builder {
      mPositiveButtonText = positiveButtonText
      return this
    }

    fun setPositiveButtonText(
      @StringRes resId: Int
    ): Builder {
      mPositiveButtonText = mHelper.context!!.getString(resId)
      return this
    }

    fun setNegativeButtonText(negativeButtonText: String?): Builder {
      mNegativeButtonText = negativeButtonText
      return this
    }

    fun setNegativeButtonText(
      @StringRes resId: Int
    ): Builder {
      mNegativeButtonText = mHelper.context?.getString(resId)
      return this
    }

    fun setTheme(@StyleRes theme: Int): Builder {
      mTheme = theme
      return this
    }

    fun build(): PermissionRequest {
      if (mRationale == null) {
        mRationale =
          mHelper.context!!.getString(string.rationale_ask)
      }
      if (mPositiveButtonText == null) {
        mPositiveButtonText = mHelper.context?.getString(R.string.ok)
      }
      if (mNegativeButtonText == null) {
        mNegativeButtonText = mHelper.context?.getString(R.string.cancel)
      }
      return PermissionRequest(
        mHelper,
        mPerms,
        mRequestCode,
        mRationale!!,
        mPositiveButtonText!!,
        mNegativeButtonText!!,
        mTheme
      )
    }
  }

  init {
    mPerms = perms.clone()
    this.requestCode = requestCode
    this.rationale = rationale
    this.positiveButtonText = positiveButtonText
    this.negativeButtonText = negativeButtonText
    this.theme = theme
  }
}